profiles/api/oauth2.py

```diff
@@ -0,0 +1,121 @@
+from json import JSONDecodeError, dumps
+from urllib.parse import urlencode
+import requests
+from flask import Blueprint, current_app, redirect, request, jsonify, url_for, json
+from werkzeug.exceptions import BadRequest
+
+OAUTH2_BP = Blueprint('oauth', __name__)
+
+
+def find_client_by_id(client_id):
+    for name, details in current_app.config['config']['oauth2']['clients'].items():
+        if details['id'] == client_id:
+            details['name'] = name
+            return details
+
+    raise KeyError
+
+
+@OAUTH2_BP.route('/authorize')
+def authorize():
+    if 'client_id' not in request.args:
+        raise BadRequest('Invalid client_id')
+
+    try:
+        client = find_client_by_id(request.args.get('client_id'))
+    except KeyError:
+        raise BadRequest('Invalid client_id')
+
+    if request.args.get('redirect_uri') != client['redirect_uri']:
+        raise BadRequest('Invalid redirect_uri')
+
+    state = {
+        'redirect_uri': client['redirect_uri'],
+        'client_id': client['id'],
+        'original': request.args.get('state', '')
+    }
```

Comment:
Rather than having to pass all this, might be better to store locally and just include the key. Also gives CSRF protection for free (if random).